core/vmx.c

```diff
@@ -296,3 +297,25 @@ void get_interruption_info_t(interruption_info_t *info, uint8_t v, uint8_t t)
     info->reserved = 0;
     info->valid = 1;
 }
+
+#define COMP_PENDING_0(name)
+#define COMP_PENDING_1(name) \
+    if (vcpu->vmx.vmcs_cache_w.name##_dirty) \
+        vmwrite(vcpu, name, vcpu->vmx.vmcs.name##_value);
+#define COMP_PENDING(cache_r, cache_w, width, name) \
+    COMP_PENDING_##cache_w(name)
+
+void vcpu_handle_vmcs_pending(struct vcpu_t* vcpu)
```

Comment:
You have changed the semantics of this function, but not where it is called. Now that it has the responsibility of flushing all the dirty VMCS values, we should delay it as much as possible, in case `vmcs_write` gets called again after the flush but before VM entry (`asm_vmxrun`). Probably we need to draw a line somewhere in `cpu_vmx_execute` or in `cpu_vmx_run`:

```
// Cached vmcs_write() is preferred over uncached vmwrite()
vmcs_write(..);
vmcs_write(..);
...
// *** No cached vmcs_write allowed after this point ***
vmwrite(..);
vmwrite(..);
...
// VM entry
```